cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerRestartService.java

```diff
@@ -0,0 +1,114 @@
+package com.sequenceiq.cloudbreak.cm;
+
+import static com.sequenceiq.cloudbreak.api.endpoint.v4.common.Status.UPDATE_IN_PROGRESS;
+import static com.sequenceiq.cloudbreak.cm.DataView.SUMMARY;
+import static com.sequenceiq.cloudbreak.event.ResourceEvent.CLUSTER_CM_CLUSTER_SERVICES_RESTARTING;
+
+import java.math.BigDecimal;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.cloudera.api.swagger.ClustersResourceApi;
+import com.cloudera.api.swagger.ServicesResourceApi;
+import com.cloudera.api.swagger.client.ApiClient;
+import com.cloudera.api.swagger.client.ApiException;
+import com.cloudera.api.swagger.model.ApiCommand;
+import com.cloudera.api.swagger.model.ApiCommandList;
+import com.cloudera.api.swagger.model.ApiRestartClusterArgs;
+import com.cloudera.api.swagger.model.ApiRolesToInclude;
+import com.cloudera.api.swagger.model.ApiRollingRestartClusterArgs;
+import com.cloudera.api.swagger.model.ApiService;
+import com.sequenceiq.cloudbreak.cm.client.retry.ClouderaManagerApiFactory;
+import com.sequenceiq.cloudbreak.cm.polling.ClouderaManagerPollingServiceProvider;
+import com.sequenceiq.cloudbreak.cm.polling.PollingResultErrorHandler;
+import com.sequenceiq.cloudbreak.dto.StackDtoDelegate;
+import com.sequenceiq.cloudbreak.polling.ExtendedPollingResult;
+import com.sequenceiq.cloudbreak.service.CloudbreakException;
+import com.sequenceiq.cloudbreak.structuredevent.event.CloudbreakEventService;
+
+@Component
+public class ClouderaManagerRestartService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClouderaManagerRestartService.class);
+
+    @Inject
+    private ClouderaManagerPollingServiceProvider clouderaManagerPollingServiceProvider;
+
+    @Inject
+    private PollingResultErrorHandler pollingResultErrorHandler;
+
+    @Inject
+    private CloudbreakEventService eventService;
+
+    @Inject
+    private ClouderaManagerApiFactory clouderaManagerApiFactory;
+
+    public int doRestartServicesIfNeeded(ApiClient apiClient, StackDtoDelegate stack, boolean waitForCommandExecutionOnly, boolean rollingRestartEnabled)
```

Comment:
I have a feeling this method is doing too much and thus we can't name it properly.
maybe having a separate method for waiting and removing the `waitForCommandExecutionOnly` flag here would help
this method could invoke the waiter one, you could have a separate method with `waitforalreadyrunnningrestart` or similar